### ISSUE TITLE
Bazel considers 502 Bad Gateway HTTP responses as unrecoverable

### ISSUE THREAD

**Issue by criemen** [bazelbuild/bazel#23334]
### Description of the bug:

GitHub downloads are unfortunately not entirely stable.
I'm getting the following spurious error on CI:
```
WARNING: Download from https://github.com/bazelbuild/rules_rust/releases/download/0.49.1/rules_rust-v0.49.1.tar.gz failed: class com.google.devtools.build.lib.bazel.repository.downloader.UnrecoverableHttpException GET returned 502 Bad Gateway
```
which results in a hard build failure:
```
INFO: Repository rules_rust~ instantiated at:
  <builtin>: in <toplevel>
Repository rule http_archive defined at:
  /home/runner/.cache/bazel/_bazel_runner/9f4ed60bba64c867c2242caf607221b5/external/bazel_tools/tools/build_defs/repo/http.bzl:382:31: in <toplevel>
Error: ERROR: An error occurred during the fetch of repository 'rules_rust~':
   Traceback (most recent call last):
	File "/home/runner/.cache/bazel/_bazel_runner/9f4ed60bba64c867c2242caf607221b5/external/bazel_tools/tools/build_defs/repo/http.bzl", line 131, column 45, in _http_archive_impl
		download_info = ctx.download_and_extract(
Error: Error in download_and_extract: java.io.IOException: Error downloading [https://github.com/bazelbuild/rules_rust/releases/download/0.49.1/rules_rust-v0.49.1.tar.gz] to /home/runner/.cache/bazel/_bazel_runner/9f4ed60bba64c867c2242caf607221b5/external/rules_rust~/temp7115974534031201847/rules_rust-v0.49.1.tar.gz: GET returned 502 Bad Gateway
ERROR: @rules_rust//rust/settings:experimental_use_cc_common_link :: Error loading option @rules_rust//rust/settings:experimental_use_cc_common_link: java.io.IOException: Error downloading [https://github.com/bazelbuild/rules_rust/releases/download/0.49.1/rules_rust-v0.49.1.tar.gz] to /home/runner/.cache/bazel/_bazel_runner/9f4ed60bba64c867c2242caf607221b5/external/rules_rust~/temp7115974534031201847/rules_rust-v0.49.1.tar.gz: GET returned 502 Bad Gateway
```

Looking at the [source code](https://github.com/bazelbuild/bazel/blob/c299b3927042c62bfdfa2888a6aea10ed396f2e8/src/main/java/com/google/devtools/build/lib/bazel/repository/downloader/HttpConnector.java#L208), bazel is not retrying downloads on error 502.
I think that's a mistake, and retrying a 502 could recover from this - as you can try out yourself, usually GitHub is serving the rules_rust release just fine at the specified URL most of the time.

### Which category does this issue belong to?

External Dependency

### Which operating system are you running Bazel on?

Linux, ubuntu 22.04

### What is the output of `bazel info release`?

release 7.2.1

**Comment by keith** [comment]
++, we see this a lot from github, probably ~daily. I didn't think to check this but seems like we should remove it from that opt out list

**Comment by meteorcloudy** [comment]
SG, can you send a PR to fix this? It should be quite easy!

**Comment by criemen** [comment]
will do!

**Comment by criemen** [comment]
I also talked to the internal team responsible for this service, and they're investigating why we send the 502 in the first place.

PR for now is at https://github.com/bazelbuild/bazel/pull/23378.

**Comment by iancha1992** [comment]
A fix for this issue has been included in [Bazel 7.4.0 RC1](https://github.com/bazelbuild/bazel/releases/tag/7.4.0rc1). Please test out the release candidate and report any issues as soon as possible.
If you're using Bazelisk, you can point to the latest RC by setting `USE_BAZEL_VERSION=7.4.0rc1`. Thanks!

**Comment by palfrey** [comment]
I'm seeing this again with 8.2.1. See https://github.com/TraceMachina/nativelink/actions/runs/25104722217/job/73562924048?pr=2267
